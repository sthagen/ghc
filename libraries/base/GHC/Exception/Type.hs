{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE BangPatterns #-}
{-# OPTIONS_HADDOCK not-home #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Exception.Type
-- Copyright   :  (c) The University of Glasgow, 1998-2002
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- Exceptions and exception-handling functions.
--
-----------------------------------------------------------------------------

module GHC.Exception.Type
       ( Exception(..)    -- Class
       , SomeExceptionWithLocation(..), SomeException(..), ArithException(..)
        , addBacktrace
       , divZeroException, overflowException, ratioZeroDenomException
       , underflowException
       , traceException
       ) where

import Data.Maybe
import Data.Typeable (Typeable, cast)
   -- loop: Data.Typeable -> GHC.Err -> GHC.Exception
import GHC.Base
import GHC.Show
import GHC.Exception.Backtrace (Backtrace)

-- TODO: Remove!
import {-# SOURCE #-} Debug.Trace

{- |
The @SomeExceptionWithLocation@ type is the root of the exception type hierarchy.
When an exception of type @e@ is thrown, behind the scenes it is
encapsulated in a @SomeException@ with is wrapped by @SomeExceptionWithLocation@.
-}
data SomeExceptionWithLocation = SomeExceptionWithLocation !SomeException ![Backtrace]

data SomeException = forall e . Exception e => SomeException !e

-- | @since 3.0
instance Show SomeExceptionWithLocation where
    -- TODO: Print backtraces
    showsPrec p (SomeExceptionWithLocation (SomeException e) _) = showsPrec p e

instance Show SomeException where
    -- TODO: Fix show
    showsPrec p (SomeException e) = showsPrec p e

-- | Add a 'Backtrace' to the list of backtraces.
--
-- @since 4.16.0.0
addBacktrace :: Backtrace -> SomeExceptionWithLocation -> SomeExceptionWithLocation
addBacktrace bt (SomeExceptionWithLocation e bts) =
    SomeExceptionWithLocation e (bt : bts)

{- |
Any type that you wish to throw or catch as an exception must be an
instance of the @Exception@ class. The simplest case is a new exception
type directly below the root:

> data MyException = ThisException | ThatException
>     deriving Show
>
> instance Exception MyException

The default method definitions in the @Exception@ class do what we need
in this case. You can now throw and catch @ThisException@ and
@ThatException@ as exceptions:

@
*Main> throw ThisException \`catch\` \\e -> putStrLn (\"Caught \" ++ show (e :: MyException))
Caught ThisException
@

In more complicated examples, you may wish to define a whole hierarchy
of exceptions:

> ---------------------------------------------------------------------
> -- Make the root exception type for all the exceptions in a compiler
>
> data SomeCompilerException = forall e . Exception e => SomeCompilerException e
>
> instance Show SomeCompilerException where
>     show (SomeCompilerException e) = show e
>
> instance Exception SomeCompilerException
>
> compilerExceptionToException :: Exception e => e -> SomeExceptionWithLocation
> compilerExceptionToException = toException . SomeCompilerException
>
> compilerExceptionFromException :: Exception e => SomeExceptionWithLocation -> Maybe e
> compilerExceptionFromException x = do
>     SomeCompilerException a <- fromException x
>     cast a
>
> ---------------------------------------------------------------------
> -- Make a subhierarchy for exceptions in the frontend of the compiler
>
> data SomeFrontendException = forall e . Exception e => SomeFrontendException e
>
> instance Show SomeFrontendException where
>     show (SomeFrontendException e) = show e
>
> instance Exception SomeFrontendException where
>     toException = compilerExceptionToException
>     fromException = compilerExceptionFromException
>
> frontendExceptionToException :: Exception e => e -> SomeExceptionWithLocation
> frontendExceptionToException = toException . SomeFrontendException
>
> frontendExceptionFromException :: Exception e => SomeExceptionWithLocation -> Maybe e
> frontendExceptionFromException x = do
>     SomeFrontendException a <- fromException x
>     cast a
>
> ---------------------------------------------------------------------
> -- Make an exception type for a particular frontend compiler exception
>
> data MismatchedParentheses = MismatchedParentheses
>     deriving Show
>
> instance Exception MismatchedParentheses where
>     toException   = frontendExceptionToException
>     fromException = frontendExceptionFromException

We can now catch a @MismatchedParentheses@ exception as
@MismatchedParentheses@, @SomeFrontendException@ or
@SomeCompilerException@, but not other types, e.g. @IOException@:

@
*Main> throw MismatchedParentheses \`catch\` \\e -> putStrLn (\"Caught \" ++ show (e :: MismatchedParentheses))
Caught MismatchedParentheses
*Main> throw MismatchedParentheses \`catch\` \\e -> putStrLn (\"Caught \" ++ show (e :: SomeFrontendException))
Caught MismatchedParentheses
*Main> throw MismatchedParentheses \`catch\` \\e -> putStrLn (\"Caught \" ++ show (e :: SomeCompilerException))
Caught MismatchedParentheses
*Main> throw MismatchedParentheses \`catch\` \\e -> putStrLn (\"Caught \" ++ show (e :: IOException))
*** Exception: MismatchedParentheses
@

-}
class (Typeable e, Show e) => Exception e where
    toException   :: e -> SomeExceptionWithLocation
    fromException :: SomeExceptionWithLocation -> Maybe e

    toException e = SomeExceptionWithLocation (SomeException e) []
    fromException (SomeExceptionWithLocation (SomeException e) _) = cast e

    -- | Render this exception value in a human-friendly manner.
    --
    -- Default implementation: @'show'@.
    --
    -- @since 4.8.0.0
    displayException :: e -> String
    displayException = show

    toString :: e -> String
    toString e = show e

instance Exception SomeException where
  toException e = SomeExceptionWithLocation e []
  fromException (SomeExceptionWithLocation e _) = Just e
  toString (SomeException e) = "SomeException " ++ toString e

instance Exception SomeExceptionWithLocation where
    toException se = se
    fromException = Just
    displayException (SomeExceptionWithLocation e _) = displayException e
    toString (SomeExceptionWithLocation e _) = "SomeExceptionWithLocation " ++ toString e

traceException :: Exception e => e -> e
traceException e = trace (toString e) e

-- |Arithmetic exceptions.
data ArithException
  = Overflow
  | Underflow
  | LossOfPrecision
  | DivideByZero
  | Denormal
  | RatioZeroDenominator -- ^ @since 4.6.0.0
  deriving ( Eq  -- ^ @since 3.0
           , Ord -- ^ @since 3.0
           )

divZeroException, overflowException, ratioZeroDenomException, underflowException  :: SomeExceptionWithLocation
divZeroException        = toException DivideByZero
overflowException       = toException Overflow
ratioZeroDenomException = toException RatioZeroDenominator
underflowException      = toException Underflow

-- | @since 4.0.0.0
instance Exception ArithException

-- | @since 4.0.0.0
instance Show ArithException where
  showsPrec _ Overflow        = showString "arithmetic overflow"
  showsPrec _ Underflow       = showString "arithmetic underflow"
  showsPrec _ LossOfPrecision = showString "loss of precision"
  showsPrec _ DivideByZero    = showString "divide by zero"
  showsPrec _ Denormal        = showString "denormal"
  showsPrec _ RatioZeroDenominator = showString "Ratio has zero denominator"
