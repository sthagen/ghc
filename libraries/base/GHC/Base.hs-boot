{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Base (Maybe, Semigroup, Monoid, String) where

import GHC.Maybe (Maybe)
import GHC.Types (Char)

class Semigroup a
class Monoid a

type String = [Char]
