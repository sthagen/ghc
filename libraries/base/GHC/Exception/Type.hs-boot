{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Exception.Type
  ( SomeExceptionWithLocation
  , divZeroException
  , overflowException
  , ratioZeroDenomException
  , underflowException
  , Exception
  ) where

import GHC.Num.Integer ()   -- See Note [Depend on GHC.Num.Integer] in GHC.Base

data SomeExceptionWithLocation
divZeroException, overflowException,
  ratioZeroDenomException, underflowException :: SomeExceptionWithLocation

class Exception e

instance Exception SomeExceptionWithLocation
