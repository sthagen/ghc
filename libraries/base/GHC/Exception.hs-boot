{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE ExistentialQuantification#-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TypeInType #-}

{-
This SOURCE-imported hs-boot module cuts a big dependency loop:

         GHC.Exception
imports  Data.Maybe
imports  GHC.Base
imports  GHC.Err
imports  {-# SOURCE #-} GHC.Exception

More dramatically

         GHC.Exception
imports  Data.Typeable
imports  Data.Typeable.Internal
imports  GHC.Arr (fingerprint representation etc)
imports  GHC.Real
imports  {-# SOURCE #-} GHC.Exception

However, GHC.Exceptions loop-breaking exports are all nice,
well-behaved, non-bottom values.  The clients use 'raise#'
to get a visibly-bottom value.
-}

module GHC.Exception
  ( module GHC.Exception.Type
  , errorCallException
  , errorCallWithCallStackException
  , throw
  ) where

import {-# SOURCE #-} GHC.Exception.Type
import GHC.Types ( Char, RuntimeRep, TYPE )
import GHC.Stack.Types ( CallStack, HasCallStack )

errorCallException :: [Char] -> SomeExceptionWithLocation
errorCallWithCallStackException :: [Char] -> CallStack -> SomeExceptionWithLocation

throw :: HasCallStack => forall (r :: RuntimeRep). forall (a :: TYPE r). forall e. Exception e => e -> a
