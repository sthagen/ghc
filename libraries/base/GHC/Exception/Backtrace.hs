{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# OPTIONS_HADDOCK not-home #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Exception.Backtrace
-- Copyright   :  (c) The GHC Team, 2020-2021
-- Authors     :  Ben Gamari, David Eichmann, Sven Tennie
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- Collect Exception backtraces with several mechanisms.
-----------------------------------------------------------------------------

module GHC.Exception.Backtrace
  ( Backtrace (..),
    BacktraceMechanism (..),
    setDefaultBacktraceMechanisms,
    getDefaultBacktraceMechanisms,
    showBacktraces,
    collectBacktraces,
  )
where

import Data.List
import Data.Maybe
import GHC.Base
import {-# SOURCE #-} GHC.ExecutionStack (Location, getStackTrace)
import {-# SOURCE #-} GHC.ExecutionStack.Internal (showStackFrames)
import GHC.IO.Unsafe
import {-# SOURCE #-} GHC.IORef
import GHC.Ptr
import GHC.Show
import {-# SOURCE #-} GHC.Stack
import {-# SOURCE #-} GHC.Stack.CCS
import {-# SOURCE #-} GHC.Stack.CloneStack

-- | An exception backtrace.
--
-- @since 4.15
data Backtrace
  = -- | a cost center profiler backtrace
    CostCenterBacktrace (Ptr GHC.Stack.CCS.CostCentreStack)
  | -- | a stack from 'GHC.Stack.HasCallStack'
    HasCallStackBacktrace GHC.Stack.CallStack
  | -- | a stack unwinding (e.g. DWARF) backtrace
    ExecutionBacktrace [GHC.ExecutionStack.Location]
  | -- | a backtrace from Info Table Provenance Entries
    IPEBacktrace [StackEntry]

-- | @since 4.15
instance Show Backtrace where
  -- TODO
  showsPrec p (CostCenterBacktrace ccs) = showsPrec p ccs
  showsPrec p (HasCallStackBacktrace ccs) = showsPrec p ccs
  showsPrec _ (ExecutionBacktrace ccs) = showStackFrames ccs
  showsPrec p (IPEBacktrace stackEntries) = showsPrec p stackEntries

-- | How to collect a backtrace when an exception is thrown.
data BacktraceMechanism
  = -- | collect a cost center stacktrace (only available when built with profiling)
    CostCenterBacktraceMech
  | -- | use execution stack unwinding with given limit
    ExecutionStackBacktraceMech (Maybe Int)
  | -- | collect backtraces from Info Table Provenance Entries
    IPEBacktraceMech
  deriving (Eq, Show)

showBacktraces :: [Backtrace] -> String
showBacktraces bts = unlines $ intersperse "" $ map show bts

currentBacktraceMechanisms :: IORef [BacktraceMechanism]
currentBacktraceMechanisms = unsafePerformIO $ newIORef []
{-# NOINLINE currentBacktraceMechanisms #-}

-- | Set how 'Control.Exception.throwIO', et al. collect backtraces.
setDefaultBacktraceMechanisms :: [BacktraceMechanism] -> IO ()
setDefaultBacktraceMechanisms = writeIORef currentBacktraceMechanisms

-- | Returns the currently selected 'BacktraceMechanism'.
getDefaultBacktraceMechanisms :: IO [BacktraceMechanism]
getDefaultBacktraceMechanisms = readIORef currentBacktraceMechanisms

-- | Collect a list of 'Backtrace' via all current default 'BacktraceMechanism'.
-- See 'setDefaultBacktraceMechanisms'
collectBacktraces :: IO [Backtrace]
collectBacktraces = do
  mech <- getDefaultBacktraceMechanisms
  catMaybes `fmap` mapM collectBacktraces' mech

-- | Collect a 'Backtrace' via the given 'BacktraceMechanism'.
collectBacktraces' :: BacktraceMechanism -> IO (Maybe Backtrace)
collectBacktraces' CostCenterBacktraceMech = do
  ptr <- getCurrentCCS ()
  return $ if ptr == nullPtr then Nothing else Just (CostCenterBacktrace ptr)
collectBacktraces' (ExecutionStackBacktraceMech _) = fmap ExecutionBacktrace `fmap` getStackTrace
collectBacktraces' IPEBacktraceMech = do
  stack <- cloneMyStack
  stackEntries <- decode stack
  return $ Just $ IPEBacktrace stackEntries
