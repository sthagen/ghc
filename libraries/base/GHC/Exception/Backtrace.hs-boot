{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Exception.Backtrace
    ( Backtrace
    , showBacktraces
    , collectBacktraces
    ) where

import GHC.Base

data Backtrace

showBacktraces :: [Backtrace] -> String

collectBacktraces :: IO [Backtrace]
