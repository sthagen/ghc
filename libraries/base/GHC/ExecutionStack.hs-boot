{-# LANGUAGE NoImplicitPrelude #-}

module GHC.ExecutionStack
  ( Location,
    getStackTrace,
  )
where

import GHC.Base
import {-# SOURCE #-} GHC.ExecutionStack.Internal (Location)

getStackTrace :: IO (Maybe [Location])
