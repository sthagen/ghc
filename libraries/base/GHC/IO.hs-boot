{-# LANGUAGE Unsafe #-}
{-# LANGUAGE NoImplicitPrelude #-}

module GHC.IO(
    IO(..),
    mplusIO,
    throwIOUserError
) where

import GHC.Types
import {-# SOURCE #-} GHC.Base (String)

mplusIO :: IO a -> IO a -> IO a
throwIOUserError :: String -> IO a
