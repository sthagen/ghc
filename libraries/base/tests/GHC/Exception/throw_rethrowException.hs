module Main where

import Control.Exception
import GHC.Exception
import GHC.Exception.Backtrace
import System.Exit

data CustomException = CustomException deriving (Show)

instance Exception CustomException

-- When an exception is thrown more than one times, it keeps the first backtrace.
main :: IO ()
main = do
  setDefaultBacktraceMechanisms [HasCallStackBacktraceMech]
  catch
    ( catch
        -- Throw for the first time.
        (throw CustomException)
        -- Throw for the second time.
        (\(e :: SomeExceptionWithLocation) -> throw e)
    )
    ( \(e :: SomeExceptionWithLocation) -> case e of
        (SomeExceptionWithLocation _ bts) ->
          case head bts of
            -- Only print the most significant location; i.e. the location
            -- where throw was called.
            HasCallStackBacktrace cs -> print $ last $ getCallStack cs
            _ -> exitFailure
    )
