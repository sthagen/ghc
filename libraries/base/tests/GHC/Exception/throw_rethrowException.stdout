("throw",SrcLoc {srcLocPackage = "main", srcLocModule = "Main", srcLocFile = "throw_rethrowException.hs", srcLocStartLine = 19, srcLocStartCol = 10, srcLocEndLine = 19, srcLocEndCol = 15})
